presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java

```diff
@@ -399,6 +772,107 @@ public void partitionPage(Page page)
             flush(false);
         }
 
+        public void ariaPartitionPage(Page page)
+        {
+            populatePositionsForEachPartition(page);
+
+            decodePage(page);
+
+            for (int i = 0; i < partitionData.length; i++) {
+                PartitionData target = partitionData[i];
+                target.appendRows(decodedBlocks, containsNonFixedWidthColumns, rowSizeForfixedWidthBlocks, variableWidthBlockChannels, outputBuffer);
+            }
+        }
+
+        private void calculateRowSizeForFixedWidthBlocks()
+        {
+            rowSizeForfixedWidthBlocks = 0;
+            containsNonFixedWidthColumns = false;
+            for (int i = 0; i < sourceTypes.size(); i++) {
+                Type type = sourceTypes.get(i);
+                Integer serializedBytesPerRowForCurrentBlock = FIXED_WIDTH_TYPE_SERIALIZED_BYTES.getOrDefault(type.getClass(), 0);
+                rowSizeForfixedWidthBlocks += serializedBytesPerRowForCurrentBlock;
+
+                if (serializedBytesPerRowForCurrentBlock == 0) {
+                    containsNonFixedWidthColumns = true;
+                    variableWidthBlockChannels.add(i);
+                }
+            }
+        }
+
+        private void decodePage(Page page)
+        {
+            int channelCount = page.getChannelCount();
+
+            if (decodedBlocks == null) {
+                decodedBlocks = new DecodedObjectNode[channelCount];
+            }
+
+            for (int i = 0; i < page.getChannelCount(); i++) {
+                decodedBlocks[i] = decodeBlock(page.getBlock(i));
+            }
+        }
+
+        // Dict/RLE is a separate node. There can't be consequetive Dict/RLE nodes because we call flatten
+        private DecodedObjectNode decodeBlock(Block block)
+        {
+            Block decodedBlock = flattener.flatten(block).get();
+
+            if (decodedBlock instanceof ArrayBlock) {
```

Comment:
It seems to me that DecodedBlockNode can be a simple POJO that combines a decoded block with its children. Making it so simplifies the code a bit:

```
        private DecodedBlockNode decodeBlock(Block block)
        {
            try (BlockLease lease = flattener.flatten(block)) {
                Block decodedBlock = lease.get();

                if (decodedBlock instanceof ArrayBlock) {
                    ColumnarArray columnarArray = ColumnarArray.toColumnarArray(decodedBlock);
                    return new DecodedBlockNode(columnarArray, ImmutableList.of(decodeBlock(columnarArray.getElementsBlock())));
                }

                if (decodedBlock instanceof MapBlock) {
                    ColumnarMap columnarMap = ColumnarMap.toColumnarMap(decodedBlock);
                    return new DecodedBlockNode(columnarMap, ImmutableList.of(decodeBlock(columnarMap.getKeysBlock()), decodeBlock(columnarMap.getValuesBlock())));
                }

                if (decodedBlock instanceof RowBlock) {
                    ColumnarRow columnarRow = ColumnarRow.toColumnarRow(decodedBlock);
                    ImmutableList.Builder<DecodedBlockNode> children = ImmutableList.builder();
                    for (int i = 0; i < columnarRow.getFieldCount(); i++) {
                        children.add(decodeBlock(columnarRow.getField(i)));
                    }
                    return new DecodedBlockNode(columnarRow, children.build());
                }

                if (decodedBlock instanceof DictionaryBlock) {
                    return new DecodedBlockNode(decodedBlock, ImmutableList.of(decodeBlock(((DictionaryBlock) decodedBlock).getDictionary())));
                }

                if (decodedBlock instanceof RunLengthEncodedBlock) {
                    return new DecodedBlockNode(decodedBlock, ImmutableList.of(decodeBlock(((RunLengthEncodedBlock) decodedBlock).getValue())));
                }

                return new DecodedBlockNode(decodedBlock, ImmutableList.of());
            }
        }

    static class DecodedBlockNode
    {
        private final Object decodedBlock;
        private final List<DecodedBlockNode> children;

        public DecodedBlockNode(Object decodedBlock, List<DecodedBlockNode> children)
        {
            this.decodedBlock = requireNonNull(decodedBlock, "decodedBlock is null");
            this.children = requireNonNull(children, "children is null");
        }

        public Object getDecodedBlock()
        {
            return decodedBlock;
        }

        public List<DecodedBlockNode> getChildren()
        {
            return children;
        }
    }
```